### PR TITLE
Add support for TS-style random tile variants.

### PR DESCRIPTION
**Pull request by pchote**
Adds more cheap polish for our TS and RA2 mods, and fixes the last major unhandled item on #5350.

### CODE REVIEW

**Comment by Mailaender** [comment]
:+1: This is very nice polish. 

You will need to address the compiler/style issues the CI is mourning about though.

**Comment by pchote** [comment]
Fixed.

To see the difference, compare [old](https://cloud.githubusercontent.com/assets/167819/6882297/6c0db110-d577-11e4-8dd2-4f2240460588.png) with [new](https://cloud.githubusercontent.com/assets/167819/6882694/4136e8e4-d586-11e4-9173-c0e7e8522eec.png).

File: OpenRA.Game/Map/TileSet.cs [hunk 1]
```diff
@@ -65,10 +65,10 @@ public class TerrainTypeInfo
 
 	public class TerrainTemplateInfo
 	{
-		static readonly TerrainTemplateInfo Default = new TerrainTemplateInfo(0, null, int2.Zero, null);
+		static readonly TerrainTemplateInfo Default = new TerrainTemplateInfo(0, new string[] { null }, int2.Zero, null);
 
 		public readonly ushort Id;
-		public readonly string Image;
+		public readonly string[] Image;
```
Comment (Phrohdoh) [inline OpenRA.Game/Map/TileSet.cs]: `Images`

**Comment by pchote** [comment]
This appears to regress TD (and probably the other mods).

**Comment by Mailaender** [comment]
Crashes the RA mod:

```
Exception of type `OpenRA.Graphics.SheetOverflowException`: Terrain sheet overflow. Try increasing the tileset SheetSize parameter.
  at OpenRA.Graphics.Theater+<Theater>c__AnonStorey0.<>m__0 () [0x00044] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Graphics/Theater.cs:53 
  at OpenRA.Graphics.SheetBuilder.Allocate (Size imageSize, float2 spriteOffset) [0x000d3] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Graphics/SheetBuilder.cs:121 
  at OpenRA.Graphics.SheetBuilder.Add (System.Byte[] src, Size size, float2 spriteOffset) [0x0003a] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Graphics/SheetBuilder.cs:71 
  at OpenRA.Graphics.SheetBuilder.Add (ISpriteFrame frame) [0x00014] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Graphics/SheetBuilder.cs:63 
  at OpenRA.Graphics.Theater+<Theater>c__AnonStorey1.<>m__1 (ISpriteFrame f) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Graphics/Theater.cs:67 
  at System.Linq.Enumerable+<CreateSelectIterator>c__Iterator10`2[OpenRA.Graphics.ISpriteFrame,OpenRA.Graphics.Sprite].MoveNext () [0x00059] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/System.Core/System.Linq/Enumerable.cs:2270 
  at System.Linq.Enumerable.Count[Sprite] (IEnumerable`1 source) [0x0002c] in /home/abuild/rpmbuild/BUILD/mono-3.12.1/mcs/class/System.Core/System.Linq/Enumerable.cs:641 
  at OpenRA.Graphics.Theater..ctor (OpenRA.TileSet tileset) [0x001be] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Graphics/Theater.cs:76 
  at OpenRA.Graphics.WorldRenderer..ctor (OpenRA.World world) [0x000c3] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Graphics/WorldRenderer.cs:63 
  at OpenRA.Game.StartGame (System.String mapUID, WorldType type) [0x000ad] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:156 
  at OpenRA.Game.LoadShellMap () [0x00019] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:380 
  at OpenRA.Mods.Common.LoadScreens.BlankLoadScreen.StartGame (OpenRA.Arguments args) [0x0022e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/LoadScreens/BlankLoadScreen.cs:97 
  at OpenRA.Game.InitializeMod (System.String mod, OpenRA.Arguments args) [0x003d7] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:372 
  at OpenRA.Mods.Common.Widgets.Logic.ModBrowserLogic+<LoadMod>c__AnonStorey1.<>m__0 () [0x00022] in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:159 
  at (wrapper delegate-invoke) <Module>:invoke_void ()
  at OpenRA.Primitives.ActionQueue.PerformActions () [0x00099] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Primitives/ActionQueue.cs:43 
  at OpenRA.Game.LogicTick () [0x00006] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:474 
  at OpenRA.Game.Loop () [0x000d4] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:598 
  at OpenRA.Game.Run () [0x00042] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:638 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:111 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:39
```

**Comment by reaperrr** [comment]
Out of couriosity I just added ```SheetSize: 1024``` to the RA desert, snow and temperate tileset definitions and that fixed the above crash.
The question: is that increased requirement an expected behaviour and @pchote just forgot to check these tilesets, or is there actually an underlying bug that increases the ```SheetSize``` more than necessary.

**Comment by pchote** [comment]
> or is there actually an underlying bug that increases the `SheetSize` more than necessary.

There was a bug.  I was stashing the result of `frames.Select(f => sheetBuilder.Add(f))` as an IEnumerable, which meant it was invoked a second time (adding a second set of frames) when it was later queried.  Storing the intermediate sprites as an array fixes the problem.

**Comment by Mailaender** [comment]
I can confirm that the crashes are gone.

**Comment by obrakmann** [comment]
At least for RA, water tiles don't get randomized anymore in the editor.

**Comment by pchote** [comment]
@obrakmann are you sure that is a new regression? This doesn't touch anything related to the legacy PickAny behaviour.

Note: I still plan on addressing @Phrohdoh's request.

**Comment by obrakmann** [comment]
You're right, happens on bleed, too. Well, that sucks.

**Comment by pchote** [comment]
Renamed `Image` &rarr; `Images` as requested by @Phrohdoh.

**Comment by obrakmann** [comment]
:+1: didn't spot any regressions in-game. Also suddenly the TS blank map doesn't look like crap anymore :D

**Comment by reaperrr** [comment]
> You're right, happens on bleed, too. Well, that sucks.

I checked the ra tilesets, there is no water with PickAny defined, so it's not a bug.

**Comment by Mailaender** [comment]
The map editor also relies on the setup of PickAny definitions to randomize tiles.

**Pull request merged**